### PR TITLE
[CDF-23599] 😮‍💨 No redeploy unnecessarily

### PR DESCRIPTION
**Pull request by doctrino**
# Description
Extend testing to ensure that we do not redeploy any resources that has not changed. 

## Checklist

- [ ] Tests added/updated.
- [ ] Run Demo Job Locally.
- [ ] Documentation updated.
- [ ] Changelogs updated in [CHANGELOG.cdf-tk.md](https://github.com/cognitedata/toolkit/blob/main/CHANGELOG.cdf-tk.md).
- [ ] Template changelogs updated in [CHANGELOG.templates.md](https://github.com/cognitedata/toolkit/blob/main/CHANGELOG.templates.md).
- [ ] Version bumped.
  [_version.py](https://github.com/cognitedata/toolkit/blob/main/cognite/cognite_toolkit/_version.py) and
  [pyproject.toml](https://github.com/cognitedata/toolkit/blob/main/pyproject.toml) per [semantic versioning](https://semver.org/).

### CODE REVIEW

File: cognite_toolkit/_cdf_tk/loaders/_resource_loaders/data_organization_loaders.py [hunk 1]
```diff
@@ -94,16 +94,31 @@ def dump_id(cls, id: str) -> dict[str, Any]:
...
-                resource["metadata"][key] = json.dumps(value)
-        if resource.get("writeProtected") is None:
-            # Todo: Setting missing default value, bug in SDK.
-            resource["writeProtected"] = False
-        if resource.get("metadata") is None:
-            # Todo: Wrongly set to empty dict, bug in SDK.
-            resource["metadata"] = {}
+            for key, value in list(resource["metadata"].items()):
+                if isinstance(value, dict):
+                    resource["metadata"][key] = json.dumps(value)
```
Comment (doctrino) [inline cognite_toolkit/_cdf_tk/loaders/_resource_loaders/data_organization_loaders.py:99]: This is a bit surprising. You can use objects in the metadata that is automatically made into a json string.

File: cognite_toolkit/_cdf_tk/loaders/_resource_loaders/data_organization_loaders.py [hunk 2]
```diff
@@ -94,16 +94,31 @@ def dump_id(cls, id: str) -> dict[str, Any]:
...
+            # Default value is empty dict, so we don't need to dump it.
+            dumped.pop("metadata")
+        if "metadata" in dumped and "metadata" in local:
+            meta_local = local["metadata"]
+            for key, value in list(dumped["metadata"].items()):
+                if isinstance(meta_local.get(key), dict | list):
+                    try:
+                        converted = json.loads(value)
+                    except json.JSONDecodeError:
+                        continue
```
Comment (doctrino) [inline cognite_toolkit/_cdf_tk/loaders/_resource_loaders/data_organization_loaders.py:117]: Added the reverse operation as well

File: tests/test_integration/test_commands/test_deploy.py [hunk 3]
```diff
@@ -42,6 +50,9 @@ def test_deploy_complete_org(cdf_tool_config: CDFToolConfig, build_dir: Path) ->
         verbose=True,
     )
 
+    changed_resources = get_changed_resources(cdf_tool_config, build_dir)
+    assert not changed_resources, "Redeploying the same resources should not change anything"
```
Comment (doctrino) [inline tests/test_integration/test_commands/test_deploy.py:54]: This is the extension of the testing using the `complete_org` and `complete_org_alpha` to ensure we have at least one example for each resource.

File: CHANGELOG.cdf-tk.md [hunk 4]
```diff
@@ -26,6 +26,9 @@ Changes are grouped as follows:
 
 - No more warning about missing `.env` file when running in `Google Cloud Build`.
 - When deploying a `Sequence` resource, Cognite Toolkit now replaces `dataSetExternalId` with `dataSetId`.
+- Cognite Toolkit has improved resources that have server set default values that can lead to redeploy even when
+  unchanged. THis includes `Sequences`, `Containers`, `DataSets`, `Views`, `Nodes`, `Edges`, `ExtractionPipelines`,
```
Comment (doctrino) [inline CHANGELOG.cdf-tk.md:30]:
```suggestion
  unchanged. This includes `Sequences`, `Containers`, `DataSets`, `Views`, `Nodes`, `Edges`, `ExtractionPipelines`,
```

**Review by ronpal — APPROVED** [review]
🚀

**Pull request merged**